File: ios/Classes/handler/FluwxAuthHandler.m
//
// Created by mo on 2018/8/23.
//

#import "FluwxAuthHandler.h"
#import "WXApi.h"
#import "WechatAuthSDK.h"

@implementation FluwxAuthHandler

WechatAuthSDK *_qrauth;
FlutterMethodChannel *_fluwxMethodChannel = nil;


- (instancetype)initWithRegistrar:(NSObject <FlutterPluginRegistrar> *)registrar methodChannel:(FlutterMethodChannel *)flutterMethodChannel {
    self = [super init];
    if (self) {
        _qrauth = [[WechatAuthSDK alloc] init];
        _qrauth.delegate = self;
        _fluwxMethodChannel = flutterMethodChannel;
    }

    return self;
}

- (void)handleAuth:(FlutterMethodCall *)call result:(FlutterResult)result {
    NSString *openId = call.arguments[@"openId"];

    [WXApiRequestHandler sendAuthRequestScope:call.arguments[@"scope"]
                                        State:(call.arguments[@"state"] == (id) [NSNull null]) ? nil : call.arguments[@"state"]
                                       OpenID:(openId == (id) [NSNull null]) ? nil : openId completion:^(BOOL done) {result(@(done));}];
}

- (void)authByQRCode:(FlutterMethodCall *)call result:(FlutterResult)result {
    NSString *appId = call.arguments[@"appId"];
    NSString *scope = call.arguments[@"scope"];
    NSString *nonceStr = call.arguments[@"nonceStr"];
    NSString *timeStamp = call.arguments[@"timeStamp"];
    NSString *signature = call.arguments[@"signature"];
    NSString *schemeData = (call.arguments[@"schemeData"] == (id) [NSNull null]) ? nil : call.arguments[@"schemeData"];

    BOOL done = [_qrauth Auth:appId nonceStr:nonceStr timeStamp:timeStamp scope:scope signature:signature schemeData:schemeData];
    result(@(done));
}

- (void)stopAuthByQRCode:(FlutterMethodCall *)call result:(FlutterResult)result {
    BOOL done = [_qrauth StopAuth];
    result(@(done));
}

- (void)onQrcodeScanned {
    [_fluwxMethodChannel invokeMethod:@"onQRCodeScanned" arguments:nil];
}

- (void)onAuthGotQrcode:(UIImage *)image {
    NSData *imageData = UIImagePNGRepresentation(image);
//    if (imageData == nil) {
//        imageData = UIImageJPEGRepresentation(image, 1);
//    }

    [_fluwxMethodChannel invokeMethod:@"onAuthGotQRCode" arguments:imageData];
}

- (void)onAuthFinish:(int)errCode AuthCode:(nullable NSString *)authCode {
    NSDictionary *errorCode = @{@"errCode": @(errCode)};
    NSMutableDictionary *result = [NSMutableDictionary dictionaryWithDictionary:errorCode];
    if (authCode != nil) {
        result[@"authCode"] = authCode;
    }
    [_fluwxMethodChannel invokeMethod:@"onAuthByQRCodeFinished" arguments:result];
}
@end

File: android/src/main/kotlin/com/jarvan/fluwx/handlers/FluwxAuthHandler.kt
/*
 * Copyright (C) 2020 The OpenFlutter Organization
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.jarvan.fluwx.handlers

import com.tencent.mm.opensdk.diffdev.DiffDevOAuthFactory
import com.tencent.mm.opensdk.diffdev.OAuthErrCode
import com.tencent.mm.opensdk.diffdev.OAuthListener
import com.tencent.mm.opensdk.modelmsg.SendAuth
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

internal class FluwxAuthHandler(private val methodChannel: MethodChannel) {
    //    private DiffDevOAuthFactory.getDiffDevOAuth()
    private val qrCodeAuth by lazy {
        DiffDevOAuthFactory.getDiffDevOAuth()
    }

    private val qrCodeAuthListener by lazy {
        object : OAuthListener {
            override fun onAuthFinish(p0: OAuthErrCode, authCode: String?) {
                methodChannel.invokeMethod("onAuthByQRCodeFinished", mapOf(
                        "errCode" to p0.code,
                        "authCode" to authCode
                ))
            }

            override fun onAuthGotQrcode(p0: String?, p1: ByteArray) {
                methodChannel.invokeMethod("onAuthGotQRCode", mapOf("errCode" to 0, "qrCode" to p1))
            }

            override fun onQrcodeScanned() {
                methodChannel.invokeMethod("onQRCodeScanned", mapOf("errCode" to 0))
            }

        }
    }


    fun sendAuth(call: MethodCall, result: MethodChannel.Result) {
        val req = SendAuth.Req()
        req.scope = call.argument("scope")
        req.state = call.argument("state")
        val openId = call.argument<String?>("openId")
        if (!openId.isNullOrBlank()) {
            req.openId = call.argument("openId")
        }

        result.success(WXAPiHandler.wxApi?.sendReq(req))
    }


    fun authByQRCode(call: MethodCall, result: MethodChannel.Result) {
        val appId = call.argument("appId") ?: ""
        val scope = call.argument("scope") ?: ""
        val nonceStr = call.argument("nonceStr") ?: ""
        val timeStamp = call.argument("timeStamp") ?: ""
        val signature = call.argument("signature") ?: ""
//        val schemeData = call.argument("schemeData")?:""

        result.success(qrCodeAuth.auth(appId, scope, nonceStr, timeStamp, signature, qrCodeAuthListener))
    }

    fun stopAuthByQRCode(result: MethodChannel.Result) {
        result.success(qrCodeAuth.stopAuth())
    }

    fun removeAllListeners() {
        qrCodeAuth.removeAllListeners()
    }
}